File: numfact_v2.py



# get all the primes less than 1000000
def getPrime():
    primes = []
    # initialize the array
    arr = [0] * 1000001
    for i in range(2, 1000001):
        if arr[i] != 1:
            # set all the multiples of prime to 1
            j = i * 2
            while j <= 1000000:
                arr[j] = 1
                j += i
    for i in range(2, len(arr)):
        # indexes with value zero are the primes
        if arr[i] == 0:
            primes.append(i)
    return primes


# calculate the factors


def populateFactorDict(factorDict, primes, number):
    primesLen = len(primes)
    tmp = number
    for i in range(primesLen):
        # if n is divisible by x increase the value at key == x by 1
        if primes[i] > tmp:
            break

        while number % primes[i] == 0 and number >= primes[i] :
            if primes[i] in factorDict:
                factorDict[primes[i]] += 1
            else:
                factorDict[primes[i]] = 1
            number /= primes[i]

    return factorDict


def getFactorMap(primes, numbers, n):
    factorDict = dict()

    for i in range(n):
        factorDict = populateFactorDict(factorDict, primes, numbers[i])
    return factorDict


def main():

    # calculate primes
    primes = getPrime()
    test = int(input())
    for _ in range(test):
        n = int(input())

        numbers = list(map(int, input().split()))

        factorDict = getFactorMap(primes, numbers, n)
        answer = 1
        for key, value in factorDict.items():
            # number of factor => 12 ->[(2,2), (3,1)] --> (2+1)*(1+1)
            answer *= value + 1
        print(answer)


main()

File: NumfactV2.java

import java.util.*;
import java.lang.*;

public class NumfactV2 {

    public static ArrayList<Integer> getPrime(){
        int [] primes = new int[1000001];
        for(int i = 2;i<1000001;i++){
            if(primes[i] ==0){
                for(int j = i*2;j<=1000000;j+=i ){
                    primes[j] = 1;
                }
            }
        }
        ArrayList<Integer>list = new ArrayList<Integer>();
        int primeLen = primes.length;
        for(int i = 2;i<primeLen;i++){
            if(primes[i] == 0){
                list.add(i);
            }
        }
        return  list;
    }

    public static HashMap<Integer, Integer>getFactorMap(ArrayList<Integer> primes, int[] numbers, int n){
        HashMap<Integer, Integer> factorDict = new HashMap<Integer, Integer>();
        for(int i =0;i<n;i++){
            factorDict = populateFactorDict(factorDict, primes,numbers[i]);
        }
        return  factorDict;
    }



    public static HashMap<Integer, Integer> populateFactorDict(HashMap<Integer, Integer>factorDict, ArrayList<Integer>primes, int  number){
        int primesLen = primes.size();
        int tmp = number;
        for(int i = 0;i<primesLen;i++){
            int prime =primes.get(i);
            if(prime>tmp) break;

            while (number%prime==0 && number>= prime){
                if(factorDict.containsKey(prime)){
                    factorDict.put(prime, factorDict.get(prime)+1);
                }
                else{
                    factorDict.put(prime, 1);
                }
                number /= prime;
            }
        }
        return  factorDict;

    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int test = sc.nextInt();
        for(int t = 0;t<test;t++){
            int n = sc.nextInt();
            int a[] = new int[n];
            for(int i =0;i<n;i++) a[i] = sc.nextInt();


            HashMap<Integer, Integer>factorDict = getFactorMap(getPrime(), a, n);
            long result = 1;

            // Getting an iterator
            Iterator hmIterator = factorDict.entrySet().iterator();
            while (hmIterator.hasNext()) {
                Map.Entry mapElement = (Map.Entry)hmIterator.next();
                String freq = mapElement.getValue().toString();
                int numberFactors = Integer.parseInt(freq);

                result *= ((long)numberFactors+1);
                //System.out.println("tmp "+result);
            }
            System.out.println(result);


        }
    }
}
